ycmd/completers/language_server/lsapi.py

```diff
@@ -144,6 +144,18 @@ def Rename( request_id, request_data, new_name ):
   } )
 
 
+def References( request_id, request_data ):
+  return BuildRequest( request_id, 'textDocument/references', {
```

Comment:
The protocol defines this as a `TextDocumentPositionParams` with an additional (mandatory) `context` object.

I also think we should include the declaration of the requested identifier too.

Perhaps we should do:

```python
request = BuildTextDocumentPositionParams( request_data )
request[ 'includeDeclaraion' ] = True
return BuildRequest( request_id, 'textDocument/references', request )
```

Reply:
Okay, I'm reading the protocol documentation and I'm not sure what `context` is supposed to be. Is the following correct?
```python
  return BuildRequest( request_id, 'textDocument/references', {
    'textDocument': {
      'uri': _MakeUriForFile( request_data[ 'filepath' ] ),
    },
    'position': {
      'line': request_data[ 'line_num' ] - 1,
      'character': request_data[ 'start_column' ] - 1,
    },
    'includeDeclaration': True,
  } )
```

Reply:
You're right i made an error. It should be:

```
  return BuildRequest( request_id, 'textDocument/references', {
    'textDocument': {
      'uri': _MakeUriForFile( request_data[ 'filepath' ] ),
    },
    'position': {
      'line': request_data[ 'line_num' ] - 1,
      'character': request_data[ 'start_column' ] - 1,
    },
    context: {
        'includeDeclaration': True,
    },
  } )
```

To explain:
* `ReferenceParams` extends `TextDocumentPositionParams` adding `context` which is an object (aka dictionary) of type `ReferenceContext`
* `ReferenceContext` is an object (aka dictionary) with a single item `includeDeclaration` of type bool

Reply:
But the point is that there is already a method for generating `TextDocumentPositionParams` part in `lsapi`

Reply:
> ReferenceParams extends TextDocumentPositionParams adding context which is an object (aka dictionary) of type ReferenceContext

So my initial understanding was actually correct. My mind just decided that it doesn't make sense to have a `dict` for only one value.

Hopefully, I've done what you expected.

Reply:
```python
request = BuildTextDocumentPositionParams( request_data )
request[ 'includeDeclaraion' ] = True
return BuildRequest( request_id, 'textDocument/references', request )
```

Wait, shouldn't this actually be:
```python
request = BuildTextDocumentPositionParams( request_data )
request[ 'context' ] = { 'includeDeclaration': True }
return BuildRequest( request_id, 'textDocument/references', request )
```

Reply:
Yes it should, that's what I was referring when I said:

> You're right i made an error